tests/unit/test_utils.cpp

```diff
@@ -44,11 +44,26 @@ VecSimQueryParams CreateQueryParams(const HNSWRuntimeParams &RuntimeParams) {
     return QueryParams;
 }
 
-void runTopKSearchTest(VecSimIndex *index, const void *query, size_t k,
+static bool is_async_index(VecSimIndex *index) {
+    return dynamic_cast<VecSimTieredIndex<float, float> *>(index) != nullptr ||
+           dynamic_cast<VecSimTieredIndex<double, double> *>(index) != nullptr;
+}
+
+void runTopKSearchTest(VecSimIndex *index, const void *query, size_t k, size_t expected_num_res,
                        std::function<void(size_t, double, size_t)> ResCB, VecSimQueryParams *params,
                        VecSimQueryResult_Order order) {
     VecSimQueryResult_List res = VecSimIndex_TopKQuery(index, query, k, params, order);
-    ASSERT_EQ(VecSimQueryResult_Len(res), k);
+    if (is_async_index(index)) {
+        // Async index may return more or less than the expected number of results,
```

Comment:
It should never return more results, no?

Reply:
if the expectation is based on the index size (which is the default expectation), we might have an out-of-date expectation (index size was 9, we looked for 10, but before/while we searched another vector was added so we found 10 results)